store/cache.go

```diff
@@ -143,11 +147,46 @@ func (cm *CacheManager) cleanup() error {
 		return nil
 	}
 
-	sort.Sort(changesByReverseMtime(fil))
-	for _, fi := range fil[cm.maxItems:] {
-		if err := os.Remove(cm.path(fi.Name())); err != nil && os.IsNotExist(err) {
-			return err
+	var lastErr error
+	sort.Sort(changesByMtime(fil))
+	deleted := 0
+	for _, fi := range fil {
+		path := cm.path(fi.Name())
+		n, err := hardLinkCount(path)
+		if err != nil {
+			logger.Noticef("cannot inspect cache: %s", err)
+		}
+		// If the file is referenced in the filesystem somewhere
+		// else our copy is "free" so skip it. If there is any
+		// error we cleanup the file (it is just a cache afterall).
+		if n > 1 {
+			continue
+		}
+		if err := osRemove(path); err != nil {
+			if !os.IsNotExist(err) {
+				logger.Noticef("cannot cleanup cache: %s", err)
+				lastErr = err
+			}
+			continue
+		}
+		deleted++
+		if len(fil)-deleted <= cm.maxItems {
+			break
+		}
+	}
+	return lastErr
+}
+
+// hardLinkCount returns the number of hardlinks for the given path
+func hardLinkCount(path string) (uint64, error) {
+	fi, err := os.Stat(path)
+	if err != nil {
+		return 0, err
+	}
+	if sys := fi.Sys(); sys != nil {
```

Comment:
couldn't this be written as `if stat, ok := fi.Sys().(*syscall.Stat_t); ok && stat != nil {`?